moveit_planners/ompl/ompl_interface/src/detail/ompl_constraints.cpp

```diff
@@ -385,4 +387,151 @@ std::shared_ptr<BaseConstraint> createOMPLConstraint(const moveit::core::RobotMo
   }
 }
-}  // namespace ompl_interface
\ No newline at end of file
+
+/******************************************
+ * Toolpath Constraint
+ * ****************************************/
+ToolPathConstraint::ToolPathConstraint(const moveit::core::RobotModelConstPtr& robot_model, const std::string& group,
+                                       unsigned int num_dofs, const unsigned int num_cons, double step)
+  : BaseConstraint(robot_model, group, num_dofs, num_cons), step_(step)
+{
+  pose_nn_.setDistanceFunction([this](const geometry_msgs::msg::Pose& pose1, const geometry_msgs::msg::Pose& pose2) {
+    return poseDistance(pose1, pose2);
+  });
+}
+
+
+void ToolPathConstraint::parseConstraintMsg(const moveit_msgs::msg::Constraints& constraints) override
+{
+  // TODO: Is this needed? Can this be done using setPath()?
+}
+
+
+void ToolPathConstraint::setPath(const moveit_msgs::msg::CartesianTrajectory& msg)
+{
+  // Clear previous path
+  pose_nn_.clear();
+
+  // Interpolate along waypoints and insert into pose_nn_;
+  for (unsigned int i = 0; i < msg.points.size() - 1; i++)
+  {
+    // Insert waypoints
+    pose_nn_.add(msg.points.at(i).point.pose);
+
+    // TODO: Consider arc lenth between the poses to determine the number of interpolation steps
+    // Linearly determine the number of steps
+    const auto num_steps = static_cast<unsigned int>(
+        euclideanDistance(msg.points.at(i).point.pose, msg.points.at(i + 1).point.pose) / step_);
+
+    // Insert interpolated poses
+    auto delta = step_;
+    for (unsigned int j = 0; j < num_steps - 1; j++)
+    {
+      const auto pose = poseInterpolator(msg.points.at(i).point.pose, msg.points.at(i + 1).point.pose, delta);
+      pose_nn_.add(pose);
+      delta += step_;
+    }
+  }
+
+  // Insert last waypoint
+  pose_nn_.add(msg.points.at(msg.points.size() - 1).point.pose);
+}
+
+void ToolPathConstraint::function(const Eigen::Ref<const Eigen::VectorXd>& joint_values,
+                                  Eigen::Ref<Eigen::VectorXd> out) const
+{
+  // TODO: try setting num_dim = 1 in BaseConstraint and using poseDistance as the constraint function
+  // Ideal: F(q_actual) - F(q_nearest_pose) = 0
+
+  // Actual end-effector pose
+  const Eigen::Isometry3d eef_pose_eigen = forwardKinematics(joint_values);
+  const Eigen::Vector3d trans = eef_pose_eigen.translation();
+  const Eigen::Quaterniond eef_quat(eef_pose_eigen.rotation());
+  const Eigen::AngleAxisd eef_aa(eef_quat);
+  const Eigen::Vector3d eef_aa_vec = eef_aa.axis() * eef_aa.angle();
+
+  geometry_msgs::msg::Pose eef_pose;
+  eef_pose.position.x = trans.x();
+  eef_pose.position.y = trans.y();
+  eef_pose.position.z = trans.z();
+
+  eef_pose.orientation.w = eef_quat.w();
+  eef_pose.orientation.x = eef_quat.x();
+  eef_pose.orientation.y = eef_quat.y();
+  eef_pose.orientation.z = eef_quat.z();
+
+  // Nearest end-effector pose in trajectory
+  const geometry_msgs::msg::Pose eef_nearest_pose = pose_nn_.nearest(eef_pose);
```

Comment:
I wonder if it makes sense to call `nearestK` instead, get the 2 nearest points, and then interpolate between those?

That would give you a nice tradeoff between having to cache a super fine set of points up front, and doing a *little* more at runtime to get something nearer to the desired pose goal.